sound/soc/sof/intel/hda-ctrl.c

```diff
@@ -161,21 +161,94 @@ int hda_dsp_ctrl_clock_power_gating(struct snd_sof_dev *sdev, bool enable)
 	return 0;
 }
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
-/*
- * While performing reset, controller may not come back properly and causing
- * issues, so recommendation is to set CGCTL.MISCBDCGE to 0 then do reset
- * (init chip) and then again set CGCTL.MISCBDCGE to 1
- */
 int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
 {
 	struct hdac_bus *bus = sof_to_bus(sdev);
-	int ret;
+	struct hdac_stream *azx_dev;
+	int ret = 0;
+
+	if (bus->chip_init)
+		return 0;
 
 	hda_dsp_ctrl_misc_clock_gating(sdev, false);
-	ret = snd_hdac_bus_init_chip(bus, full_reset);
+
+	if (full_reset) {
+		/* clear STATESTS */
+		snd_hdac_chip_writew(bus, STATESTS, STATESTS_INT_MASK);
+
+		/* reset HDA controller */
+		ret = hda_dsp_ctrl_link_reset(sdev, true);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to reset HDA controller\n");
+			return ret;
+		}
+
+		usleep_range(500, 1000);
+
+		/* exit HDA controller reset */
+		ret = hda_dsp_ctrl_link_reset(sdev, false);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to exit HDA controller reset\n");
+			return ret;
+		}
+
+		usleep_range(1000, 1200);
+	}
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	/* check to see if controller is ready */
+	if (!snd_hdac_chip_readb(bus, GCTL)) {
+		dev_dbg(bus->dev, "controller not ready!\n");
+		return -EBUSY;
+	}
+
+	/* Accept unsolicited responses */
+	snd_hdac_chip_updatel(bus, GCTL, AZX_GCTL_UNSOL, AZX_GCTL_UNSOL);
+
+	/* detect codecs */
+	if (!bus->codec_mask) {
+		bus->codec_mask = snd_hdac_chip_readw(bus, STATESTS);
+		dev_dbg(bus->dev, "codec_mask = 0x%lx\n", bus->codec_mask);
+	}
+#endif
+
+	/* clear stream status */
+	list_for_each_entry(azx_dev, &bus->stream_list, list)
+		snd_hdac_stream_writeb(azx_dev, SD_STS, SD_INT_MASK);
+
+	/* clear WAKESTS */
+	snd_hdac_chip_writew(bus, STATESTS, STATESTS_INT_MASK);
```

Comment:
Humm... I wonder what wasn't clear in my explanations.

**We shall not use snd_hdac_() calls unless it's protected by SND_SOC_SOF_HDA**

You are doing the exact opposite. Is this really intentional?

Reply:
@plbossart What I understand first was that reverse. As I talked with @keyonjie , he suggested that we can use the HDA lib head file (the defined **snd_hdac_()** functions ), but not the HDA lib C files. and he suggested that we replace all the **snd_sof_dsp_()** with the **snd_hdac_()** functions defined in the HDA lib head. 
@keyonjie Could you state you points?
If what we understand not right, I will push another version.

Reply:
@keyonjie One point is that we have not defined all the registers in SOF, but they are all defined in the HDA head files, right?

Reply:
Hi @plbossart IMHO, we should use snd_hdac_() version in all possible places, as those snd_hdac_() are all headers only, the io_ops implementation is in our hda_bus.c.

As we aligned before SOF HDA/HDMI implemented last year, we came out to that we can use hdac library header files for both I2S(BSD) and HDA/HDMI(GPL), since we have already used snd_hdac_() inside hdac library(during invoking hdac APIs), so I suggest changing to use this unified interface in our sof hda related registers read/write operations, for both I2S and HDA/HDMI cases.

@lgirdwood any comments?

Reply:
@lgirdwood